### DISCUSSION TITLE
JMX vs JMXMP

### DISCUSSION THREAD

**Discussion by grzesuav** [cryostatio/cryostat#637] [Q&A] [answered]
hi, wanted to check out if this is still the issue. - https://meteatamel.wordpress.com/2012/02/13/jmx-rmi-vs-jmxmp/

especially the part 
`
This basically tells the client to connect to a JMX server on the specified host and port over RMI. At least, that’s what you’d expect right? Well, the reality is a little different. When the client connects to host over that port, the JMX server responds with something like “Ok, I got your connect request on port1, now you can connect on port2” and accordingly to [this blog post](http://blog.markfeeney.com/2010/10/jmx-through-ssh-tunnel.html), port2 is chosen randomly.
`
I am bit away from JVM world but wanted to confirm if this is still default behaviour

**Comment by grzesuav** [comment]
https://docs.oracle.com/en/java/javase/22/jmx/jmx-technology-versions.html#GUID-5971D87D-657F-4768-BF21-04A5407CB5C8

**Comment by andrewazores** [comment, marked as answer]
For this reason, we always recommend to use single-port RMI with a known port number:

https://cryostat.io/get-started/#using-jmx

When Cryostat automatically discovers your applications by ex. Kubernetes Endpoints querying, it [always uses JMX RMI on a single port](https://github.com/cryostatio/cryostat/blob/775aee4cdd61f5b8d91aa38f4601933c9750d54e/src/main/java/io/cryostat/discovery/KubeApiDiscovery.java#L600).

You can be extra-specific and set both port number properties on your application deployment:

```
-Dcom.sun.management.jmxremote.port=1234
-Dcom.sun.management.jmxremote.rmi.port=1234
```

to the same known number, but this should have the same effect as setting the one property alone.

https://www.baeldung.com/jmx-ports

Note that the "local attach port" can be left alone and on a random port number, because that isn't involved with remote connections over the network such as ones initiated by Cryostat. So it's only the JMX and RMI ports that need to be exposed and which should be set to the same number.

**Reply by grzesuav** [reply]
got it, thanks for explaining !
